message-broker/src/lib.rs

```diff
@@ -86,6 +141,50 @@ decl_module! {
 				);
 				*up = up.split_off(num);
 			});
+
+			// Sending HRMP messages is a little bit more involved. On top of the number of messages
+			// per block limit, there is also a constraint that it's possible to send only a single
+			// message to a given recipient per candidate.
+			let mut non_empty_hrmp_channels = <Self as Store>::NonEmptyHrmpChannels::get();
+			let outbound_hrmp_num = cmp::min(HRMP_MSG_NUM_PER_CANDIDATE, non_empty_hrmp_channels.len());
+			let mut outbound_hrmp_messages = Vec::with_capacity(outbound_hrmp_num);
+			let mut prune_empty = Vec::with_capacity(outbound_hrmp_num);
+
+			for &recipient in non_empty_hrmp_channels.iter().take(outbound_hrmp_num) {
```

Comment:
Doesn't this mean, if there are less parachains we want to send a message to than the upper number of messages per channel, we will waste space?

Assume the upper number is 5, but we only send messages to 2 parachains. This means we waste some pontential message space.


And the current algorithm favors parachains with lower ids. Meaning if we always send a message to one of these parachains, we may starve one of the higher parachains.

Reply:
> And the current algorithm favors parachains with lower ids. Meaning if we always send a message to one of these parachains, we may starve one of the higher parachains.

I see that this is discussed below.

Reply:
> Doesn't this mean, if there are less parachains we want to send a message to than the upper number of messages per channel, we will waste space?

Per candidate you mean? Yes, that's correct. We cannot send more than one message in a channel per one block. This is an artificial limitation though, it is not required for HRMP. It is mainly dictated by XCMP - there it is important.

Reply:
Ahh okay.